feat(matlab)!: replace parser

acristoffers opened the pull request:
I wrote a MATLAB parser which has more features than the current one:

- It detects commands. The code
```matlab
variable P(1, 2, 3) semidefinte
```
returns a `function_call` token with 2 `argument` tokens, which is how matlab sees it. I use a scanner to correctly detect the line, so it won't mix it with a function call followed by a identifier.
- It detects string escapes and formatting. This returns a string with two tokens inside, differentiating escape sequences from formatting options. Unfortunately there is an issue with tree-sitter that prevents me from doing it with single-quoted string at the moment, but it works fine with double-quoted ones.
```matlab
"Format: \t %2.3f"
```
- It detects line continuations, lambdas, parfor statement, arguments statements (a kind of MATLAB block), two more operators, superclass function call, and when it comes to classes, it correctly detects and generates specific tokens for everything (MATLAB has some weird constructs, like the properties block).

The parser also generates more tokens, which allows to better handle not only syntax highlighting, but specially textobjects. For instance, matrices and cells are composed of rows with values, which allows the creation of textobjects to operate on those, which is something I was missing a lot. At this point I believe I have implemented everything in MATLAB's grammar.

clason commented:
* Commit message should be `feat(matlab)!: replace parser` (breaking change).

* `tags.scm` are not used in this project, so please drop that file.

* Consider adding an `injections.scm` (at least for comments, but ideally for any language -- pure C? -- where it makes sense).

* Consider adding `@spell` captures where appropriate.

The full list of allowed captures (which is different from upstream/Helix) is here: https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights.

acristoffers commented:
>    Commit message should be feat(matlab)!: replace parser (breaking change).

Fixed.

>    tags.scm are not used in this project, so please drop that file.

Removed.

>    Consider adding an injections.scm (at least for comments, but ideally for any language -- pure C? -- where it makes sense).

Was already there, only comments. MATLAB does not allow injections.

>    Consider adding @spell captures where appropriate.

Was already there, at the end of the file. For strings and comments only, as most other projects do.

> The full list of allowed captures (which is different from upstream/Helix) is here: https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#highlights.

I fixed the function call one, but it seems to be the only trully different from what I had already.

clason commented on queries/matlab/highlights.scm:
```diff
@@ -1,107 +1,216 @@
+; @text.todo         Todo
+; @comment           Comment
+; @punctuation       Delimiter
+; @constant          Constant
+; @constant.builtin  Special
+; @constant.macro    Define
+; @define            Define
+; @macro             Macro
+; @string            String
+; @string.escape     SpecialChar
+; @string.special    SpecialChar
+; @character         Character
+; @character.special SpecialChar
+; @number            Number
+; @boolean           Boolean
+; @float             Float
+; @function          Function
+; @function.builtin  Special
+; @function.macro    Macro
+; @parameter         Identifier
+; @method            Function
+; @field             Identifier
+; @property          Identifier
+; @constructor       Special
+; @conditional       Conditional
+; @repeat            Repeat
+; @label             Label
+; @operator          Operator
+; @keyword           Keyword
+; @exception         Exception
+; @variable          Identifier
+; @type              Type
+; @type.definition   Typedef
+; @storageclass      StorageClass
+; @structure         Structure
+; @namespace         Identifier
+; @include           Include
+; @preproc           PreProc
+; @debug             Debug
+; @tag               Tag
```
please remove

clason commented on lua/nvim-treesitter/parsers.lua:
```diff
@@ -926,10 +926,10 @@ list.markdown_inline = {
   install_info = {
-    url = "https://github.com/mstanciu552/tree-sitter-matlab",
-    files = { "src/parser.c" },
+    url = "https://github.com/acristoffers/tree-sitter-matlab",
```
```suggestion
    url = "https://github.com/acristoffers/tree-sitter-matlab",
    branch = "main",
```

clason replied:
Shouldn't be required if `main` is the default branch, but the download currently fails. (May just be a hiccup on GH's side, though; let's give the archive a bit of time to propagate through the CDN. Making changes in the repo and here at the same time will interfere with CI.)

clason commented:
We also have [indents](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#indents); you could try adding at least initial support for them?

clason commented:
You also need to change the commit hash in the `lockfile.json`.

acristoffers commented:
> We also have [indents](https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#indents); you could try adding at least initial support for them? (Not required.)

I forgot about it, thank you for the reminder. It's in the repo now. And I also changed the commit in the lockfile.

acristoffers commented:
Just did two small fixes to use the @boolean and @error captures.

lucario387 commented on queries/matlab/indents.scm:
```diff
@@ -0,0 +1,32 @@
+ (catch)
+ ] @indent.branch
+
+((matrix_definition (row) @indent.align)
+ (#set! indent.open_delimiter "["))
+((cell_definition (row) @indent.align)
+ (#set! indent.open_delimiter "{"))
+((parenthesized_expression) @indent.align
+ (#set! indent.open_delimiter "("))
```
Have you done some tests locally and found out that not having closing delimiter results in more correct indentation? If yes, then I'm okay with this, if not, please add their respective closing braces.

acristoffers replied:
I didn't know about the closing delimiters, I followed the code I found in this repo and it worked, so I left it like that. I'll add the `close_delimiter` now.

acristoffers replied:
I added it. Made no difference when testing, though.

acristoffers commented on queries/matlab/indents.scm:
```diff
@@ -0,0 +1,32 @@
+ (properties)
+ ] @indent.begin
+
+[(elseif_statement)
+ (else_statement)
+ (case)
+ (otherwise)
+ (catch)
+ ] @indent.branch
```
It's the entire block. I tested it as I did it, and it works. They look like

```
(else
 else: (keyword)
(block))
```

lucario387 replied:
I'd say to only have the `keyword` of `else: (keyword)` be `@indent.branch` and similarly to the other in these sections, the whole block doesn't really need it. It will look at the smallest node at the start of the line to consider for self-dedenting, and capturing such node will be better

acristoffers replied:
Thank you for clarifying how it works. I changed accordingly. Does the same hapen with `@indent.begin`? All the grammars I saw put that one in the whole tag...

lucario387 replied:
Generally, yes. But no need to change `@indent.begin`,as your current version aligns with other `@indent.begin` in other `indents.scm` file

clason commented:
Note that the `locals` captures also differ from upstream: https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#locals

lucario387 reviewed (commented):
`indents.scm` looks fine to me now :+1:

clason commented:
`@structure` query is illegal in nvim-treesitter (for the moment).

acristoffers commented:
> `@structure` query is illegal in nvim-treesitter (for the moment).

Just saw it. Already changed it to be just `@keyword`.

acristoffers commented:
I had a `calloc` in the parser which was unnecessary, so I changed the code to just use the stack instead.

acristoffers commented:
When I first wrote this I was thinking about taking screenshots of the existing grammar and mine to show the difference, and then I thought it would be too much trouble and just wrote that text. Now I just realized that Helix uses the existing one too, so I only had to open helix and neovim to compare both. Night and day.

![parser](https://github.com/nvim-treesitter/nvim-treesitter/assets/2191364/abf5b58c-1d0b-4eec-8a61-33305d150df2)

acristoffers commented:
I added textobjects to some classes constructs, which required a small grammar adaptation.

lucario387 commented on queries/matlab/highlights.scm:
```diff
@@ -1,107 +1,174 @@
+  (operator) @number
+  (number))
+
+(unary_operator
+  (operator) @operator
+  (_))
+(binary_operator
+  (operator) @operator)
+(comparison_operator
+  (operator) @operator)
+(boolean_operator
+  (operator) @operator)
+(postfix_operator
+  (operator) @operator)
+(not_operator
+  (operator) @operator)
+(metaclass_operator
+  (operator) @operator
+  (identifier) @variable)
+(handle_operator
+  (operator) @operator
```
Similarly to keywords and variables captures, too. I don't think you need to do subcapture this specific. Usually for highlights.scm you only do captures like such if it deviates a lot from the default capture for a node with such name, which you haven't set a default yet

lucario387 replied:
Just to be sure, I'm okay with your style of captures, but it comes with more burden in you to update the queries when you eventually introduces new node types that aren't listed here

acristoffers replied:
> Captures that are listed earlier (smaller linenr) has lower precedence.

Lower? I thought it had higher precedence if it appeared first, as if the search would stop once if found a match.

lucario387 replied:
It's the reverse in neovim, and according to tree-sitter's creator, they also prefer this style of approach: 
See https://github.com/nvim-treesitter/nvim-treesitter/issues/3944#issuecomment-1458782497

clason replied:
To be specific, Neovim simply applies all matches, and last one obviously wins (as they "paint over" each other). You can check this with `:Inspect`.

acristoffers replied:
Done. It really simplified things a bit.

acristoffers commented:
The change in order revealed that I had a named node and some aliases with the same name, but totally different contexts. Fixed that.

acristoffers commented:
MATLAB accepts cell-indexing as a function identifier, `a{1}(1)`. Fixed that.

acristoffers commented:
Fixed the matrix parsing, it now follows MATLAB way of doing it a bit closer.

amaanq commented:
There are some design choices and many issues with the grammar and queries that I do not agree with/are incorrect and I think are better fixed upstream before adding here. Let me know if you'd be welcome to a PR on my end, that'd be easier than writing a bunch of comments here

acristoffers commented:
> There are some design choices and many issues with the grammar and queries that I do not agree with/are incorrect and I think are better fixed upstream before adding here. Let me know if you'd be welcome to a PR on my end, that'd be easier than writing a bunch of comments here

Sure, go ahead.

amaanq commented:
https://github.com/acristoffers/tree-sitter-matlab/pull/1

acristoffers commented:
I added a file that should not be in the commit. The last force-push only removes that file.

amaanq commented:
Is the pr ready to be merged?

acristoffers commented:
Yes.

acristoffers commented:
I just created nvim-treesitter/nvim-treesitter-context#292 and nvim-treesitter/nvim-treesitter-textobjects#468 to follow this one.
